### PR TITLE
add worker cancellation

### PR DESCRIPTION
**Pull request by jrhizor**
I'm planning on adding the api endpoint after https://github.com/airbytehq/airbyte/pull/2385 is merged.

For testing I was using the exchangerates dev version here, and then running something like `docker run --rm --network=host temporalio/tctl:1.7.0 workflow cancel --workflow_id 4fd2ab70-20e7-4ebf-8670-66d93e218d49` to cancel the workflow. Once #2385 is merged we'll have metadata that lets us find this workflow id for a specific job and cancel based on the job id instead.

Blocking this merge:
- [x] remove exchangerate testing
- [x] review

### CODE REVIEW

File: airbyte-integrations/connectors/source-exchangeratesapi-singer/source_exchangeratesapi_singer/source.py
Comment (jrhizor) [inline airbyte-integrations/connectors/source-exchangeratesapi-singer/source_exchangeratesapi_singer/source.py:53]: will remove this testing code before merging

File: airbyte-workers/src/main/java/io/airbyte/workers/temporal/CancellationHandler.java
Comment (cgardens) [inline airbyte-workers/src/main/java/io/airbyte/workers/temporal/CancellationHandler.java:47]: can you explain what this line does? is the idea that from outside the worker cancel gets called via the temporal client so that the next time this heart beat happens an exception will be thrown?

**Review by cgardens — APPROVED** [review]

**Comment by cgardens** [comment]
Heads up #2385 is blocked on a UI change so likely won't happen until early next week.

File: airbyte-workers/src/main/java/io/airbyte/workers/temporal/CancellationHandler.java
Comment (jrhizor) [reply airbyte-workers/src/main/java/io/airbyte/workers/temporal/CancellationHandler.java:47]: Yeah exactly.

Comment (davinchia) [reply airbyte-workers/src/main/java/io/airbyte/workers/temporal/CancellationHandler.java:47]: can we add a comment explaining this? I would not remember the reason for this a week from now :P

File: airbyte-workers/src/main/java/io/airbyte/workers/temporal/CancellationHandler.java
Comment (sherifnada) [inline airbyte-workers/src/main/java/io/airbyte/workers/temporal/CancellationHandler.java:34]: could you add a comment explaining how to use the input `callback`?

File: airbyte-workers/src/main/java/io/airbyte/workers/temporal/CancellationHandler.java
Comment (sherifnada) [inline airbyte-workers/src/main/java/io/airbyte/workers/temporal/CancellationHandler.java:49]: why is the callback being run only being invoked if there is an exception?

File: airbyte-workers/src/main/java/io/airbyte/workers/temporal/TemporalAttemptExecution.java
Comment (sherifnada) [inline airbyte-workers/src/main/java/io/airbyte/workers/temporal/TemporalAttemptExecution.java:124]: Some questions: 
1. Why does the cancellation handler have a heartbeat method? I would expect it to have a `cancel` method. I don't follow the purpose of the heartbeat here. Is the handler a long-lived process, or is it a function that gets called on demand? 
2. Why does the heartbeat method seem to be running the actual cancellation code?

**Review by sherifnada — CHANGES_REQUESTED** [review]
I don't understand the PR well enough to give an approve so I did request changes. LMK if it's faster to sync offline.

File: airbyte-workers/src/main/java/io/airbyte/workers/DefaultSyncWorker.java
Comment (davinchia) [inline airbyte-workers/src/main/java/io/airbyte/workers/DefaultSyncWorker.java:153]: curious: what was cancelling the source/destination prior to this change?

Comment (sherifnada) [reply airbyte-workers/src/main/java/io/airbyte/workers/DefaultSyncWorker.java:153]: nothing

File: airbyte-workers/src/main/java/io/airbyte/workers/temporal/SyncWorkflow.java
Comment (davinchia) [inline airbyte-workers/src/main/java/io/airbyte/workers/temporal/SyncWorkflow.java:115]: Same thought here, except the even longer anonymous function was even more intimidating hah.

What about something like this?
```
 public StandardSyncOutput run(JobRunConfig jobRunConfig,
                                  IntegrationLauncherConfig sourceLauncherConfig,
                                  IntegrationLauncherConfig destinationLauncherConfig,
                                  StandardSyncInput syncInput)
        throws TemporalJobException {
      final var sourceLauncher = new AirbyteIntegrationLauncher(sourceLauncherConfig.getJobId(), Math.toIntExact(sourceLauncherConfig.getAttemptId()),
          sourceLauncherConfig.getDockerImage(), pbf);
      final var destinationLauncher = new AirbyteIntegrationLauncher(destinationLauncherConfig.getJobId(),
          Math.toIntExact(destinationLauncherConfig.getAttemptId()), destinationLauncherConfig.getDockerImage(), pbf);
      final var srcdockerImage = sourceLauncherConfig.getDockerImage();
      final var dstdockerImage = destinationLauncherConfig.getDockerImage();

      TemporalAttemptExecution<StandardSyncInput, StandardSyncOutput> attempt = new TemporalAttemptExecution<>(
          workspaceRoot,
          jobRunConfig,
          getSyncWorker(jobRunConfig, sourceLauncher, srcdockerImage, destinationLauncher, dstdockerImage, syncInput),
          () -> syncInput,
          new CancellationHandler.TemporalCancellationHandler());
      return attempt.get();
    }

    private CheckedFunction<Path, Worker<StandardSyncInput, StandardSyncOutput>, Exception> getSyncWorker(
                                                                                                          JobRunConfig jobRunConfig,
                                                                                                          IntegrationLauncher sourceLauncher,
                                                                                                          String sourceDockerImage,
                                                                                                          IntegrationLauncher destinationLauncher,
                                                                                                          String dstLauncherImage,
                                                                                                          StandardSyncInput syncInput) {
      return (unusedJobRoot) -> {
        // reset jobs use an empty source to induce resetting all data in destination.
        final AirbyteSource airbyteSource =
            sourceDockerImage.equals(WorkerConstants.RESET_JOB_SOURCE_DOCKER_IMAGE_STUB) ? new EmptyAirbyteSource()
                : new DefaultAirbyteSource(sourceLauncher);

        return new DefaultSyncWorker(
            jobRunConfig.getJobId(),
            Math.toIntExact(jobRunConfig.getAttemptId()),
            airbyteSource,
            new NamespacingMapper(syncInput.getPrefix()),
            new DefaultAirbyteDestination(destinationLauncher),
            new AirbyteMessageTracker(),
            NormalizationRunnerFactory.create(
                dstLauncherImage,
                pbf,
                syncInput.getDestinationConfiguration()));
      };
    }
```

File: airbyte-workers/src/test/java/io/airbyte/workers/temporal/TemporalAttemptExecutionTest.java
Comment (davinchia) [inline airbyte-workers/src/test/java/io/airbyte/workers/temporal/TemporalAttemptExecutionTest.java:79]: nit: I've been slowly introducing JUnit 5's `@DisplayName` annotation in the test classes I write. It's nice as it provides more human readable context in general. It's very helpful as test get more complicated. I don't necessary think it's needed here, since these are simple. Just letting ya know in case you haven't heard of it :)

File: airbyte-workers/src/main/java/io/airbyte/workers/temporal/TemporalAttemptExecution.java
Comment (davinchia) [inline airbyte-workers/src/main/java/io/airbyte/workers/temporal/TemporalAttemptExecution.java:105]: nit: I tend to adopt uncle bob's recommendations of writing functions that aren't too long. This method is straightforward so it's more of a stylistic thing here. Not blocking.

File: airbyte-workers/src/main/java/io/airbyte/workers/temporal/TemporalAttemptExecution.java
Comment (davinchia) [inline airbyte-workers/src/main/java/io/airbyte/workers/temporal/TemporalAttemptExecution.java:103]: I don't see this `jobRoot` argument used anywhere. Can we simplify the worker supplier signature? Am I missing something?

File: airbyte-workers/src/main/java/io/airbyte/workers/temporal/CheckConnectionWorkflow.java
Comment (davinchia) [inline airbyte-workers/src/main/java/io/airbyte/workers/temporal/CheckConnectionWorkflow.java:95]: This method is short enough it's digestable as is. Although it'll still be nice to have type information and separate the constructor call from the `.get` call. Not blocking.

File: airbyte-workers/src/main/java/io/airbyte/workers/temporal/DiscoverCatalogWorkflow.java
Comment (davinchia) [inline airbyte-workers/src/main/java/io/airbyte/workers/temporal/DiscoverCatalogWorkflow.java:97]: Same readability thoughts as `CheckConnectionWorkflow`

**Review by davinchia — CHANGES_REQUESTED** [review]
This makes sense. Most of my comments are code structure to make things more readable. Readability is of course very personal so very open for discussion. I also have some minor questions for understanding in the `TemporalAttemptExecution` class.

File: airbyte-workers/src/main/java/io/airbyte/workers/temporal/TemporalAttemptExecution.java
Comment (cgardens) [reply airbyte-workers/src/main/java/io/airbyte/workers/temporal/TemporalAttemptExecution.java:103]: this may have been an oversight on my part. if it's not used it can be removed.

File: airbyte-workers/src/main/java/io/airbyte/workers/temporal/TemporalAttemptExecution.java
Comment (cgardens) [reply airbyte-workers/src/main/java/io/airbyte/workers/temporal/TemporalAttemptExecution.java:71]: i think the objective here is to catch that _any_ exception that is thrown (checked or unchecked). It could be that it would be more coherent to remove Exception from the method signature if it not being used and instead  catch all runtime exceptions and convert them to temporal exceptions.

File: airbyte-workers/src/main/java/io/airbyte/workers/temporal/TemporalAttemptExecution.java
Comment (michel-tricot) [inline airbyte-workers/src/main/java/io/airbyte/workers/temporal/TemporalAttemptExecution.java:54]: If you call it INPUT, then you should rename T to be OUTPUT

**Review by sherifnada — APPROVED** [review]
Thanks for the refactors/context!

**Review by davinchia — APPROVED** [review]
Looks great!

**Pull request merged**
